drivers/pcie/endpoint/pcie_ep_bcm_iproc.c

```diff
@@ -0,0 +1,335 @@
+
+	/* Write offset to Configuration Indirect Address register */
+	pcie_write32(offset, &cfg->base->paxb_config_ind_addr);
+
+	/* Read data from Configuration Indirect Data register */
+	*data = pcie_read32(&cfg->base->paxb_config_ind_data);
+
+	return 0;
+}
+
+static void iproc_pcie_conf_write(struct device *dev, uint32_t offset,
+				  uint32_t data)
+{
+	const struct iproc_pcie_ep_config *cfg = dev->config_info;
+
+	/* Write offset to Configuration Indirect Address register */
+	pcie_write32(offset, &cfg->base->paxb_config_ind_addr);
+
+	/* Write data to Configuration Indirect Data register */
+	pcie_write32(data, &cfg->base->paxb_config_ind_data);
+}
+
+static int iproc_pcie_map_addr(struct device *dev, uint64_t pcie_addr,
+			       uint64_t *mapped_addr, uint32_t size,
+			       enum pcie_ob_mem_type ob_mem_type)
+{
+	const struct iproc_pcie_ep_config *cfg = dev->config_info;
+	struct iproc_pcie_ep_ctx *ctx = dev->driver_data;
+	uint64_t pcie_ob_base, pcie_ob_size, pcie_addr_start, offset;
+	uint32_t mapped_size;
+	enum pcie_outbound_map idx;
+	k_spinlock_key_t key;
+	int ret;
+
+	key = k_spin_lock(&ctx->ob_map_lock);
+
+	/* We support 2 outbound windows,
+	 * one in highmem region and another in lowmem region
+	 */
+	if ((ob_mem_type == PCIE_OB_HIGHMEM ||
+	     ob_mem_type == PCIE_OB_ANYMEM) && !ctx->highmem_in_use) {
+		idx = PCIE_MAP_HIGHMEM_IDX;
+		pcie_ob_base = cfg->map_high_base;
+		pcie_ob_size = cfg->map_high_size;
+	} else if ((ob_mem_type == PCIE_OB_LOWMEM ||
+		    ob_mem_type == PCIE_OB_ANYMEM) && !ctx->lowmem_in_use) {
+		idx = PCIE_MAP_LOWMEM_IDX;
+		pcie_ob_base = cfg->map_low_base;
+		pcie_ob_size = cfg->map_low_size;
+	} else {
+		ret = -EBUSY;
+		goto out;
+	}
+
+	/* check if the selected OB window supports size we want to map */
+	if (size > pcie_ob_size) {
+		ret = -ENOTSUP;
+		goto out;
+	}
+
+	/* Host PCIe address should be aligned to outbound window size */
+	pcie_addr_start = pcie_addr & ~(pcie_ob_size - 1);
+
+	/* Program OARR with PCIe outbound address */
+	pcie_write32(((pcie_ob_base & ~(pcie_ob_size - 1)) | PAXB_OARR_VALID),
+		     &cfg->base->paxb_oarr[idx].lower);
+	pcie_write32(pcie_ob_base >> 32, &cfg->base->paxb_oarr[idx].upper);
+
+	/* Program OMAP with Host PCIe address */
+	pcie_write32((uint32_t)pcie_addr_start,
+		     &cfg->base->paxb_omap[idx].lower);
+	pcie_write32((uint32_t)(pcie_addr_start >> 32),
+		     &cfg->base->paxb_omap[idx].upper);
+
+	/* Mark usage of outbound window */
+	if (idx == PCIE_MAP_HIGHMEM_IDX) {
+		ctx->highmem_in_use = true;
+	} else {
+		ctx->lowmem_in_use = true;
+	}
+
+	/* offset holds extra size mapped due to alignment requirement */
+	offset = pcie_addr - pcie_addr_start;
+	*mapped_addr = pcie_ob_base + offset;
+	mapped_size = pcie_ob_size - offset;
+	ret = ((mapped_size >= size) ? size : mapped_size);
+out:
+	k_spin_unlock(&ctx->ob_map_lock, key);
+
+	return ret;
+}
+
+static void iproc_pcie_unmap_addr(struct device *dev, uint64_t mapped_addr)
```

Comment:
question about data types: I see a few places where `uint64_t` is used to store memory addresses, like `mapped_addr` here and elsewhere?

Should we be using `uintptr_t`?
Also I'm interested to know whether these addresses are specifically physical and not virtual; if these are virtual addresses perhaps we could just use `void *`?

Reply:
> Should we be using `uintptr_t`?

`mapped_addr`  is PCIe outbound memory address and can be 64-bit disregard of `uintptr_t` as we have this memory in 64-bit range as well. For example, when we build zephyr for cortex-M7 , `uintptr_t` is 32-bit (32-bit wide pointer), but when we build for cortex-A72 , `uintptr_t` is 64-bit (64-bit wide pointer). In both the cases, `mapped_addr`  can actually be 64-bit wide, so  I am using `uint64_t` here.

> Also I'm interested to know whether these addresses are specifically physical and not virtual; if these are virtual addresses perhaps we could just use `void *`?

These addresses are issued by core when MMU is enabled, so these are virtual address. They are 1-to-1 physical-to-virtual address mapped. (mapping is done just to mark these addresses as nGnRnE device memory).
For the same reason as above, using `uint64_t` instead of `void *`.

Reply:
I think I understand, thanks for the extra explanation. Is it correct to say that this mechanism is for a PCI-e peripheral endpoint and these addresses are from the host CPUs perspective?

Other than my suggestion to use size_t for sizes, no objections to this PR.

Reply:
> Is it correct to say that this mechanism is for a PCI-e peripheral endpoint and these addresses are from the host CPUs perspective?

Not all addresses are from the host CPUs perspective. One of the jobs of this driver is to map the Host address to EP's local memory mapped PCIe outbound address. I have put details under public API pcie_ep_map_addr's description, where `pcie_addr` is from host CPUs perspective and `mapped_addr`(mapped outbound address) is from local CPU perspective.

> Other than my suggestion to use size_t for sizes, no objections to this PR.

I will look into this and update.